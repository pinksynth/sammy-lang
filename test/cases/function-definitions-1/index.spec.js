const {
  performSnapshotAssertions,
  expectCompiledEval,
} = require("../../helpers")
performSnapshotAssertions("Function definitions", __dirname)

test("Functions with multiple signatures", () => {
  expectCompiledEval(`
		function greet() { "Hello." }
		function greet(name) { "{greet()} My name's {name}." }
		function greet(name age) { "{greet(name)} I'm {age}." }

		[
			greet("Jim" 30)
			greet("James")
			greet()
		]
	`).toEqual([
    "Hello. My name's Jim. I'm 30.",
    "Hello. My name's James.",
    "Hello.",
  ])
})
