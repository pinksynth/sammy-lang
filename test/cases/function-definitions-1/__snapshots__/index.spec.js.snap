// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ðŸ“¸ Function definitions 1`] = `
[
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 1,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "FUNCTION",
    "value": "function",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 18,
    "columnNumberStart": 10,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "multiply",
  },
  {
    "columnNumberEnd": 19,
    "columnNumberStart": 18,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "PAREN_OPEN",
    "value": "(",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 19,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 22,
    "columnNumberStart": 21,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "b",
  },
  {
    "columnNumberEnd": 23,
    "columnNumberStart": 22,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "PAREN_CLOSE",
    "value": ")",
  },
  {
    "columnNumberEnd": 24,
    "columnNumberStart": 23,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 25,
    "columnNumberStart": 24,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "CURLY_OPEN",
    "value": "{",
  },
  {
    "columnNumberEnd": 26,
    "columnNumberStart": 25,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 27,
    "columnNumberStart": 26,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 28,
    "columnNumberStart": 27,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 29,
    "columnNumberStart": 28,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "OPERATOR_INFIX",
    "value": "*",
  },
  {
    "columnNumberEnd": 30,
    "columnNumberStart": 29,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 31,
    "columnNumberStart": 30,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "VAR",
    "value": "b",
  },
  {
    "columnNumberEnd": 32,
    "columnNumberStart": 31,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 33,
    "columnNumberStart": 32,
    "lineNumberEnd": 1,
    "lineNumberStart": 1,
    "tokenType": "CURLY_CLOSE",
    "value": "}",
  },
  {
    "columnNumberEnd": 1,
    "columnNumberStart": 33,
    "lineNumberEnd": 2,
    "lineNumberStart": 1,
    "tokenType": "WHITESPACE",
    "value": "
",
  },
  {
    "columnNumberEnd": 9,
    "columnNumberStart": 1,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "FUNCTION",
    "value": "function",
  },
  {
    "columnNumberEnd": 10,
    "columnNumberStart": 9,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 18,
    "columnNumberStart": 10,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "VAR",
    "value": "multiply",
  },
  {
    "columnNumberEnd": 19,
    "columnNumberStart": 18,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "PAREN_OPEN",
    "value": "(",
  },
  {
    "columnNumberEnd": 20,
    "columnNumberStart": 19,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 21,
    "columnNumberStart": 20,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "PAREN_CLOSE",
    "value": ")",
  },
  {
    "columnNumberEnd": 22,
    "columnNumberStart": 21,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 23,
    "columnNumberStart": 22,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "CURLY_OPEN",
    "value": "{",
  },
  {
    "columnNumberEnd": 24,
    "columnNumberStart": 23,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 32,
    "columnNumberStart": 24,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "VAR",
    "value": "multiply",
  },
  {
    "columnNumberEnd": 33,
    "columnNumberStart": 32,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "PAREN_OPEN",
    "value": "(",
  },
  {
    "columnNumberEnd": 34,
    "columnNumberStart": 33,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 35,
    "columnNumberStart": 34,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 36,
    "columnNumberStart": 35,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "VAR",
    "value": "a",
  },
  {
    "columnNumberEnd": 37,
    "columnNumberStart": 36,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "PAREN_CLOSE",
    "value": ")",
  },
  {
    "columnNumberEnd": 38,
    "columnNumberStart": 37,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "WHITESPACE",
    "value": " ",
  },
  {
    "columnNumberEnd": 39,
    "columnNumberStart": 38,
    "lineNumberEnd": 2,
    "lineNumberStart": 2,
    "tokenType": "CURLY_CLOSE",
    "value": "}",
  },
]
`;

exports[`ðŸ“¸ Function definitions 2`] = `
{
  "children": [
    {
      "args": [
        {
          "columnNumberStart": 19,
          "lineNumberStart": 1,
          "type": "IDENTIFIER",
          "value": "a",
        },
        {
          "columnNumberStart": 21,
          "lineNumberStart": 1,
          "type": "IDENTIFIER",
          "value": "b",
        },
      ],
      "children": [
        {
          "columnNumberStart": 28,
          "left": {
            "columnNumberStart": 26,
            "lineNumberStart": 1,
            "type": "IDENTIFIER",
            "value": "a",
          },
          "lineNumberStart": 1,
          "operator": "*",
          "right": {
            "columnNumberStart": 30,
            "lineNumberStart": 1,
            "type": "IDENTIFIER",
            "value": "b",
          },
          "type": "BINARY_EXPR",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 1,
      "name": "multiply",
      "type": "FUNCTION_DEFINITION",
    },
    {
      "args": [
        {
          "columnNumberStart": 19,
          "lineNumberStart": 2,
          "type": "IDENTIFIER",
          "value": "a",
        },
      ],
      "children": [
        {
          "children": [
            {
              "columnNumberStart": 33,
              "lineNumberStart": 2,
              "type": "IDENTIFIER",
              "value": "a",
            },
            {
              "columnNumberStart": 35,
              "lineNumberStart": 2,
              "type": "IDENTIFIER",
              "value": "a",
            },
          ],
          "columnNumberStart": 24,
          "function": {
            "columnNumberStart": 24,
            "lineNumberStart": 2,
            "type": "IDENTIFIER",
            "value": "multiply",
          },
          "lineNumberStart": 2,
          "type": "FUNCTION_CALL",
        },
      ],
      "columnNumberStart": 1,
      "lineNumberStart": 2,
      "name": "multiply",
      "type": "FUNCTION_DEFINITION",
    },
  ],
  "type": "ROOT",
}
`;

exports[`ðŸ“¸ Function definitions 3`] = `"function multiply$2(a,b){return a*b};function multiply$1(a){return multiply$2(a,a)}"`;
